drivers/mipi_dbi/mipi_dbi_spi.c

```diff
@@ -45,84 +49,228 @@ uint32_t var = MIPI_DBI_SPI_READ_REQUIRED;
  */
 #define MIPI_DBI_DC_BIT BIT(9)
 
-static int mipi_dbi_spi_write_helper(const struct device *dev,
+static inline int
+mipi_dbi_spi_write_helper_3wire(const struct device *dev,
+				const struct mipi_dbi_config *dbi_config,
+				bool cmd_present, uint8_t cmd,
+				const uint8_t *data_buf, size_t len)
+{
+	const struct mipi_dbi_spi_config *config = dev->config;
+	struct mipi_dbi_spi_data *data = dev->data;
+	struct spi_buf buffer;
+	struct spi_buf_set buf_set = {
+		.buffers = &buffer,
+		.count = 1,
+	};
+	int ret = 0;
+
+	/*
+	 * 9 bit word mode must be used, as the command/data bit
+	 * is stored before the data word.
+	 */
+
+	if ((dbi_config->config.operation & SPI_WORD_SIZE_MASK)
+	    != SPI_WORD_SET(9)) {
+		return -ENOTSUP;
+	}
+	buffer.buf = &data->spi_byte;
+	buffer.len = 2;
+
+	/* Send command */
+	if (cmd_present) {
+		data->spi_byte = cmd;
+		ret = spi_write(config->spi_dev, &dbi_config->config, &buf_set);
+		if (ret < 0) {
+			goto out;
+		}
+	}
+	/* Write data, byte by byte */
+	for (size_t i = 0; i < len; i++) {
+		data->spi_byte = MIPI_DBI_DC_BIT | data_buf[i];
+		ret = spi_write(config->spi_dev, &dbi_config->config, &buf_set);
+		if (ret < 0) {
+			goto out;
+		}
+	}
+out:
+	return ret;
+}
+
+static inline int
+mipi_dbi_spi_write_helper_4wire_8bit(const struct device *dev,
 				     const struct mipi_dbi_config *dbi_config,
 				     bool cmd_present, uint8_t cmd,
 				     const uint8_t *data_buf, size_t len)
 {
 	const struct mipi_dbi_spi_config *config = dev->config;
-	struct mipi_dbi_spi_data *data = dev->data;
 	struct spi_buf buffer;
 	struct spi_buf_set buf_set = {
 		.buffers = &buffer,
 		.count = 1,
 	};
 	int ret = 0;
 
-	ret = k_mutex_lock(&data->lock, K_FOREVER);
-	if (ret < 0) {
-		return ret;
+	/*
+	 * 4 wire mode is much simpler. We just toggle the
+	 * command/data GPIO to indicate if we are sending
+	 * a command or data
+	 */
+
+	buffer.buf = &cmd;
+	buffer.len = sizeof(cmd);
+
+	if (cmd_present) {
+		/* Set CD pin low for command */
+		gpio_pin_set_dt(&config->cmd_data, 0);
+		ret = spi_write(config->spi_dev, &dbi_config->config, &buf_set);
+		if (ret < 0) {
+			goto out;
+		}
 	}
 
-	if (dbi_config->mode == MIPI_DBI_MODE_SPI_3WIRE &&
-	    IS_ENABLED(CONFIG_MIPI_DBI_SPI_3WIRE)) {
-		/* 9 bit word mode must be used, as the command/data bit
-		 * is stored before the data word.
-		 */
-		if ((dbi_config->config.operation & SPI_WORD_SIZE_MASK)
-		    != SPI_WORD_SET(9)) {
-			return -ENOTSUP;
+	if (len > 0) {
+		buffer.buf = (void *)data_buf;
+		buffer.len = len;
+
+		/* Set CD pin high for data */
+		gpio_pin_set_dt(&config->cmd_data, 1);
+		ret = spi_write(config->spi_dev, &dbi_config->config, &buf_set);
+		if (ret < 0) {
+			goto out;
 		}
-		buffer.buf = &data->spi_byte;
-		buffer.len = 2;
+	}
+out:
+	return ret;
+}
+
+static inline int
+mipi_dbi_spi_write_helper_4wire_16bit(const struct device *dev,
```

Comment:
Could we add a check similar to `MIPI_DBI_SPI_READ_REQUIRED` to see if this function implementation is needed? Basically a check to see if any of the MIPI DBI SPI nodes in the devicetree use 16 bit transfers

Reply:
Done, also for the default 8 bit transfer. Because of situations where 16 bit would be the only required transfer mode, the 8 bit mode can be opt-out in same manner.